### PR TITLE
IA-2402: Org Unit Change Requests - review page

### PR DESCRIPTION
**Pull request by beygorghor**
[Mock-ups ](https://whimsical.com/georegistry-changes-requests-mobile-and-web-Qpd28e5mkQQuJcJsqjPP5N)

Page to list and approve change requests on org units

Related JIRA tickets : IA-2402

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- list page,
- filters
- dialogs
- typing
- hooks

## How to test

Create some org unit changes in you back-end  ('/admin/iaso/orgunitchangerequest/') and come back to the dashboard.

Click on org units / Review change proposals

Play with the page

## Print screen / video

Screenshot 2023-12-20 at 16 43 25
Screenshot 2023-12-20 at 16 41 31
Screenshot 2023-12-20 at 16 41 21

## Notes
Multi selection is not done yet, this is much more complex to implement

### CODE REVIEW

File: hat/assets/js/apps/Iaso/utils/permissions.ts [hunk 1]
```diff
@@ -21,6 +21,7 @@ const POLIO_SUPPLY_CHAIN_READ = 'iaso_polio_vaccine_supply_chain_read';
 const POLIO_SUPPLY_CHAIN_WRITE = 'iaso_polio_vaccine_supply_chain_write';
 const PROJECTS = 'iaso_projects';
 const REGISTRY = 'iaso_registry';
+const ORG_UNITS_CHANGE_REQUEST_APPROVE = 'iaso_org_unit_change_request_approve';
```
Comment (kemar) [inline hat/assets/js/apps/Iaso/utils/permissions.ts]: I would suggest using the term "review" instead of "approve" because you can either **approve** or **reject** a change request:

```suggestion
const ORG_UNITS_CHANGE_REQUEST_REVIEW = 'iaso_org_unit_change_request_review';
```

It's not blocking. It's up to you but it could be more explicit.

File: hat/assets/js/apps/Iaso/domains/app/translations/fr.json [hunk 2]
```diff
@@ -42,6 +42,16 @@
     "iaso.attachments.uploadError": "Erreur de chargement du fichier",
     "iaso.attachments.uploadSuccess": "Fichier envoyé avec succès",
     "iaso.CDS": "Centre de Santé",
+    "iaso.changeRequest.addComment": "Ajouter un commentaire pour le rejet",
+    "iaso.changeRequest.closingDate": "Date de fermeture",
+    "iaso.changeRequest.comment": "Commentaire",
+    "iaso.changeRequest.edit": "Validate selected requests",
```
Comment (kemar) [inline hat/assets/js/apps/Iaso/domains/app/translations/fr.json]:
```suggestion
    "iaso.changeRequest.edit": "Valider les changements sélectionnés",
```

File: hat/assets/js/apps/Iaso/domains/app/translations/fr.json [hunk 3]
```diff
@@ -278,6 +288,8 @@
     "iaso.forms.type": "Type",
     "iaso.forms.update": "Mise à jour du formulaire",
     "iaso.forms.updated_at": "Modification",
+    "iaso.forms.updated_by": "Updated by",
```
Comment (kemar) [inline hat/assets/js/apps/Iaso/domains/app/translations/fr.json]:
```suggestion
    "iaso.forms.updated_by": "Mis à jour par",
```

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/messages.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        id: 'iaso.label.created_by',
+        defaultMessage: 'Created by',
+    },
+    updated_by: {
+        id: 'iaso.forms.updated_by',
+        defaultMessage: 'Updated by',
+    },
+    multiSelectionAction: {
+        defaultMessage: 'Validate selected requests',
+        id: 'iaso.changeRequest.edit',
```
Comment (kemar) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/messages.ts]: `defaultMessage` and `id` seems to have a different meaning?

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    newFields,
+    isNew,
+    submitChangeRequest,
+}) => {
+    const { formatMessage } = useSafeIntl();
+    const [isRejectDialogOpen, setIsRejectDialogOpen] =
+        useState<boolean>(false);
+    const [rejectedReason, setRejectedReason] = useState<string | undefined>();
+    const selectedFields = newFields.filter(field => field.isSelected);
+    const handlConfirm = useCallback(() => {
```
Comment (kemar) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx]:
```suggestion
    const handelConfirm = useCallback(() => {
```

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                                color="error"
+                                autoFocus
+                            >
+                                {formatMessage(MESSAGES.reject)}
+                            </Button>
+                        </Box>
+                        <Box pl={1} display="inline-block">
+                            <Button
+                                data-test="confirm-button"
+                                onClick={handlConfirm}
```
Comment (kemar) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx]:
```suggestion
                                onClick={handleConfirm}
```

**Review by kemar — COMMENTED** [review]
Almost LGTM but there are a few missing French translations and typos.

I will soon update the API doc and I may fix my feedback myself at the same time.

I will then approve this PR.

File: hat/assets/js/apps/Iaso/utils/permissions.ts (hunk 1 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/utils/permissions.ts:24]: "Approve selected changes"

File: hat/assets/js/apps/Iaso/domains/app/translations/fr.json (hunk 2 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/app/translations/fr.json]: "Approve selected changes"

File: hat/assets/js/apps/Iaso/constants/routes.js [hunk 7]
```diff
@@ -473,6 +474,34 @@ export const orgUnitsDetailsPath = {
     ],
 };
 
+export const orgUnitChangeRequestPath = {
+    baseUrl: baseUrls.orgUnitsChangeRequest,
+    permissions: [Permission.ORG_UNITS_CHANGE_REQUEST_APPROVE],
+    component: props => <ApproveOrgUnitChanges {...props} />,
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/constants/routes.js]: `ReviewOrgUnitChanges` would be better I think (nitpick)

File: hat/assets/js/apps/Iaso/domains/app/translations/en.json [hunk 8]
```diff
@@ -42,6 +42,16 @@
     "iaso.attachments.uploadError": "Error uploading file",
     "iaso.attachments.uploadSuccess": "File uploaded successfully",
     "iaso.CDS": "Health centre",
+    "iaso.changeRequest.addComment": "Add comment for rejection",
+    "iaso.changeRequest.closingDate": "Closing date",
+    "iaso.changeRequest.comment": "Comment",
+    "iaso.changeRequest.edit": "Validate selected requests",
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/app/translations/en.json]: "review"

File: hat/assets/js/apps/Iaso/domains/app/translations/en.json [hunk 9]
```diff
@@ -278,6 +288,8 @@
     "iaso.forms.type": "Type",
     "iaso.forms.update": "Update form",
     "iaso.forms.updated_at": "Updated",
+    "iaso.forms.updated_by": "Updated by",
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/app/translations/en.json:290]: It seems strange that we don't already have a translation for updated by somewhere 🤷‍♂️
Anyway, it would make sense to use a more generic label than `iaso.forms` (nitpick)

File: hat/assets/js/apps/Iaso/domains/app/translations/en.json [hunk 10]
```diff
@@ -42,6 +42,16 @@
...
+    "iaso.changeRequest.addComment": "Add comment for rejection",
+    "iaso.changeRequest.closingDate": "Closing date",
+    "iaso.changeRequest.comment": "Comment",
+    "iaso.changeRequest.edit": "Validate selected requests",
+    "iaso.changeRequest.newValue": "Modified version",
+    "iaso.changeRequest.oldValue": "Initial version",
+    "iaso.changeRequest.openingDate": "Opening date",
+    "iaso.changeRequest.reject": "Reject all",
+    "iaso.changeRequest.seeApprovedChanges": "See approved change(s)",
+    "iaso.changeRequest.seeRejectedChanges": "See rejected change(s)",
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/app/translations/en.json:53]: Not sure it makes much sense to have both titles. The interface is pretty clear about whether changes were approved or not

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx (hunk 5 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx]: lol

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        submitChangeRequest({
+            status: 'approved',
+            approved_fields: selectedFields.map(field => field.key),
+        });
+    }, [selectedFields, submitChangeRequest]);
+    const handleReject = useCallback(() => {
+        setIsRejectDialogOpen(false);
+        submitChangeRequest({
+            status: 'rejected',
+            approved_fields: [],
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx]: Funny that you have to pass a value to `approved_fields`, I would expect the API to just take the status and a comment

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+            rejection_comment: rejectedReason,
+        });
+    }, [rejectedReason, submitChangeRequest]);
+    return (
+        <>
+            <SimpleModal
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx]: ❤️

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+                                autoFocus
+                                disabled={
+                                    !rejectedReason ||
+                                    rejectedReason.length === 0
+                                }
+                            >
+                                {formatMessage(MESSAGES.confirm)}
+                            </Button>
+                        </Box>
+                    </>
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesButtons.tsx]: That's a component

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesDialog.tsx [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+                                    >
+                                        {formatMessage(MESSAGES.selection)}
+                                    </Box>
+                                </TableCell>
+                            )}
+                        </TableRow>
+                    </TableHead>
+                    <TableBody>
+                        {newFields.map(field => {
+                            const isCellRejected =
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesDialog.tsx]: That would be a sub-component as well

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesDialog.tsx [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+}) => {
+    const { formatMessage } = useSafeIntl();
+    const handleClick = useCallback(() => {
+        setSelectedChangeRequest({
+            id: changeRequestId,
+            index,
+        });
+    }, [changeRequestId, index, setSelectedChangeRequest]);
+    return (
+        <IconButtonMui
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/ApproveOrgUnitChangesDialog.tsx]: Why not the IconButton from bluesquare-components?

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/Table/ApproveOrgUnitChangesTable.tsx [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+            Header: formatMessage(MESSAGES.created_by),
+            id: 'created_by__username',
+            accessor: 'created_by',
+            Cell: ({
+                value: createdBy,
+            }: {
+                value: NestedUser;
+            }): ReactElement => (
+                <>{createdBy ? getDisplayName(createdBy) : '--'}</>
+            ),
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/Table/ApproveOrgUnitChangesTable.tsx]: this would probably be useful in a table utils

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/Table/ApproveOrgUnitChangesTable.tsx [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+    const handleCloseDialog = useCallback(() => {
+        setSelectedChangeRequest(undefined);
+    }, []);
+    return (
+        <>
+            <ApproveOrgUnitChangesDialog
+                isOpen={Boolean(selectedChangeRequest)}
+                selectedChangeRequest={selectedChangeRequest}
+                closeDialog={handleCloseDialog}
+            />
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/Table/ApproveOrgUnitChangesTable.tsx]: I would put that in the table columns. Also, you won't be able to deep link this way.
To me it would make sense to have a similar setup as polio campaigns:
- a "list" API wit only the data needed for the list view (lighter payload)
- a "details" API that is called when opening the modal and receives the id. That's how we deeplink the campaigns modal in polio.
cf what I'm trying to do [here](https://github.com/BLSQ/iaso/pull/1002)

But it's mostly that I don't like having the modal here, because I like the jsx to "mimic" the interface as close as possible.

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/hooks/useNewFields.tsx [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+    NestedOrgUnitType,
+    NestedLocation,
+    OrgUnitChangeRequestDetails,
+} from '../types';
+import { MarkerMap } from '../../../../components/maps/MarkerMapComponent';
+import { LinkToOrgUnit } from '../../components/LinkToOrgUnit';
+import { ShortOrgUnit } from '../../types/orgUnit';
+import MESSAGES from '../messages';
+
+export type NewField = {
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/hooks/useNewFields.tsx]: `NewOrgUnitField` will age better if we have new fields elsewhere in the future

File: hat/assets/js/apps/Iaso/utils/permissions.ts (hunk 1 above)
Comment (quang-le) [reply hat/assets/js/apps/Iaso/utils/permissions.ts]: I agree

**Review by quang-le — CHANGES_REQUESTED** [review]
### TL;DR
- Blocking for immediate release:
    - Unmodified fields appearing green in Approved changes
- Blocking for short term:
    - Renaming statuses
    - updating "initial status" when a change request for the same irg unit has been approved
- Blocking in medium term (before doing multi select):
    - Moving modal from main page to table (or update front-end guidelines and bring other pages in line)
    - Add missing tooltips on icon buttons
    - Solve the problem of the intercation with Org unit's history tab (which is able to roll back the changes).
    


### Full review

About the interface:
- Statuses: I think we should use other terms 
    - "New" and "Rejected" can confuse with org unit validation status. I propose "Pending" and "Dismissed"
    - "Approved" is confusing as well, since reviewers can choose to approve only part of the request. I propose "Processed".
 - Approved requests:
     - In screenshot below, only the opening date was changed, but the parents and groups are highlighted in green as well, even though there was no difference between the original and the request
     
Screenshot 2024-01-03 at 11 40 43

- When reviewing requests for the same org unit: if the user reviews and approves changes, then reviews the second change request, the initial org unit will not have been updated. It seems the interface shows the diff between the "initial org unit" and the changed org unit at the time of the request.
- Translations:
   -  Replace "Validate" and "Validation" with "Approve" and "Approval" (the english use of "validate" is not same as the french)
   - In modal, replace "validate or reject changes". with "review changes":
   
Screenshot 2024-01-03 at 11 47 35

- Approved changes can still be rolled back via the org unit history tab. Such rollbacks will not be visible in the change request interface.

- Out of scope but related: the groups dropdown list is full of duplicates that can't be told apart:
Screenshot 2024-01-03 at 11 59 22

- No tooltip
Screenshot 2024-01-03 at 12 07 35
s on table icon buttons?

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/hooks/useNewFields.tsx [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+const getGroupsValue = (groups: NestedGroup[]): ReactElement | string => {
+    return groups && groups.length > 0
+        ? groups.map(group => group.name).join(', ')
+        : textPlaceholder;
+};
+
+const getLocationValue = (location: NestedLocation): ReactElement => {
+    return (
+        <MarkerMap
+            longitude={location.latitude}
```
Comment (madewulf) [inline hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/hooks/useNewFields.tsx]: You inverted longitude and latitude ;-)

File: hat/assets/js/apps/Iaso/domains/app/translations/en.json (hunk 10 above)
Comment (beygorghor) [reply hat/assets/js/apps/Iaso/domains/app/translations/en.json:53]: i'm using it in tooltip now

File: hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/Table/ApproveOrgUnitChangesTable.tsx (hunk 17 above)
Comment (beygorghor) [reply hat/assets/js/apps/Iaso/domains/orgUnits/approveChanges/Table/ApproveOrgUnitChangesTable.tsx]: I will let it like that, this is working and we still don't know how to manage multiple reviews

**Review by quang-le — APPROVED** [review]
Good for v1
